waveorder/waveorder_reconstructor.py

```diff
@@ -977,23 +988,40 @@ def Stokes_recon(self, I_meas):
         Parameters
         ----------
             I_meas        : numpy.ndarray
-                            polarization-sensitive intensity images with the size of (N_channel, ...)
+                            polarization-sensitive intensity images with the size of (N_channel, N, M) or
+                            (N_channel, N, M, N_defocus)
                           
         Returns
         -------
             S_image_recon : numpy.ndarray
-                            reconstructed Stokes parameters with the size of (N_Stokes, ...)
+                            reconstructed Stokes parameters with the size of (N_Stokes, N, M) or
+                            (N_channel, N, M, N_defocus)
                        
                               
         '''
-        
-        img_shape = I_meas.shape
-        
-        A_pinv = np.linalg.pinv(self.A_matrix)
-        S_image_recon = np.reshape(np.dot(A_pinv, I_meas.reshape((self.N_channel, -1))), (self.N_Stokes,)+img_shape[1:])
 
-            
-        return S_image_recon
+        n_dims = I_meas.ndim
+        if n_dims not in (3, 4):
+            raise ValueError('Unsupported image data size. Image data must be of size (N_channel, N, M) or '
+                             '(N_channel, N, M, N_defocus)')
```

Comment:
In uPTI reconstruction pipeline, my intensity data is usually structured as (N_channel, N_pattern, N, M) for 2D imaging and (N_channel, N_pattern, N, M, N_defocus) for `Stokes_recon` because the code is used to support N-D processing. 

Could we keep the original implementation for N-D processing and conditionally have the instrument-tensor processing activated when `self.A_matrix_inv` has `ndim = 4`?

Reply:
Thanks @lihaoyeh, I'll rework this implementation. I also saw that in recOrder @camFoltz passes XYZT stacks to this function, and that throws an error too. 

@lihaoyeh can you please include a test for uPTI data?

Reply:
That's a good idea. I can work on an integrated test for uPTI simulation similar to what's in `waveorder/tests/reconstructor/test_2D_QLIPP_recon.py`. The current bottleneck is to figure out how to run this test with minimum computing resources because the full forward/inverse simulation for uPTI can be computationally intense. Before I finish that, you can temporarily run the notebooks in `waveorder/examples/uPTI_simulation/` for testing. Those notebooks should at least provide the check on the array dimensions. We will have to rely on the review, for now, to identify whether there could be potential accuracy issues.

Reply:
> Thanks @lihaoyeh, I'll rework this implementation. I also saw that in recOrder @camFoltz passes XYZT stacks to this function, and that throws an error too.

@ieivanov do you mean XYZC here?  I do not ever pass an array with a time dimension XYZT or XYZCT into any of these functions, my only use case is at the level of image volumes.